### ISSUE TITLE
Innacurate buy prices due to DUST... Affects sale price for whole order... Potential DCA's and losses can result...

### ISSUE THREAD

**Issue by TheBigShort111** [taniman/profit-trailer#1293]
# Yet another case for a control to allow or exclude DUST from buy/sell orders...

![screenshot 2018-02-27 12 03 02](https://user-images.githubusercontent.com/36313988/36744906-003bce20-1bbc-11e8-983e-8b64edd15c92.png)
![screenshot 2018-02-27 12 03 17](https://user-images.githubusercontent.com/36313988/36744920-0a69b006-1bbc-11e8-80f9-a36cb2b3b5b9.png)

In this example, an order was placed for 0.001804 BTC, but the buy price and qty in the sales log does not match the buy price and qty on the exchange... The reason is that $.03 (0.000003) worth of DUST was added into the $19.05 order, changing the bought price of the whole order to $10,565.00, instead of the actual $10,558.48 that is reflected on the exchange.

The problem is twofold: First, the ability to accurately account for the individual order profit becomes ambiguous because the tiny $.03 fraction of the entire order throws off the true base price of the transaction. Second, it affects the price at which this whole order has to sell to be profitable and in certain circumstances could mean the difference between a straightforward profitable sale or perhaps a non-sale resulting in a DCA bag. Once I reset the correct price, the order went from -1.30 to +.20 or simply put, from red to green. At a false indicated -1.3% (because of the dust) this could trigger DCA trailing for some people, when in actuality 99.84% of the order is technically in profit of .20% - So in simple terms, in this example three cents could make the difference between a straght-forward, non-DCA'd sale at a potentially higher profit, or a DCA'd order that will double the exposure and potentially reduce the overall profit.

It does not make sense that such a tiny fraction (0.0016%, or roughly three cents worth) the bought price is granted the weight of determining the base price of 99.84% of the entire order. This happens so frequently that is really creates issues in accurately accounting for a true profit not only on individual orders, but on the total profit earned over time.

Surely anyone that understands what is really happening when DUST is figured into the mix would elect to turn it off, given the ability.

A simple control/switch/argument would seem to work.
Enable_SELL_DUST="true/false"
or something to that nature.

I really think this would help resolve additional related issues that people are having with DUST affecting their orders.

Thanks in advance for considering a workaround for this.

**Comment by YoJeff** [comment]
This is not as straightforward as you think.    

What about a coin that rises above dust?  What is dust?  Who determines dust?   Why would you only want to trade a portion of a position?   Once you buy a coin, you realize the exchange doesn't segregate dust -vs- non-dust?  It's just coin.   If the average cost is wrong, that's a separate issue.

The only 'official' record that matters is the exchange.   The sales log is transient.

> Surely anyone that understands what is really happening when DUST is figured into the mix would elect to turn it off, given the ability.

You presume too much.   So if we run with your assumption, you will always accumulate dust... Until it's not dust anymore -- then we are back to square one.

**Comment by TheBigShort111** [comment]
Please read the following in it's entirety...

With all respect (sincerely) - Please consider the following...

Dust exists on the exchange because the exchange rounds trades which creates the dust in the first place. This factually creates inherent problems with PT doing the accounting, while the exchange does it efficiently and as expected. On the exchange, (in a manual trade) YOU the owner of the account, are buying "x" amount of that coin at a specific price (whether you place a market order or a limit order) For this example, let's say you elect to buy BTC with 100% of your available BTC (which includes previous dust) at $10,000 even. This order will be placed for EXACTLY $10,000 (as long as the order book can fill it in one shot)

At this point, the insignificant DUST has been added regardless if it is at profit or loss, either way, it is factored into the order at a bought price of $10,000, right along with the rest of the order. Not only is this understandable as the DUST is such an insignificant part of the order, but it makes the trade very straight-forward. This method is used (and preferred) because the DUST is the most insignificant part of the equation. PT on the other hand does the opposite, it treats the DUST bought price as the base for the whole order, which creates a host issues ranging from trades sold at a loss and misreported as a profit, to possible DCA's, etc...

As mentioned, "the only'"official' record that matters is the exchange" - I agree completely, however, PT should be able to reflect the same record in the sales log as that on the exchange, all the while presenting it's users and accurate profit percentage gain (or loss) on such trades, AND preventing possible DCA's as a result from the current DUST incorporation model.

And while a fix may not be easy "per se" (I'm not a programmer) all I 'm saying is that having a control for the end user would be a simple task for the end user to enable or disable.

However, that said, let's consider another approach, how about PT treat DUST the same way the exchange does and set/reset the bought price for DUST to the order buy price set by PT (the way the exchange does), and let the DUST be the part of the trade to adopt the insignificant gain or loss adjustment instead of the whole order. This is how it's treated on the exchange. It would keep the record on PT in line with that of the exchange as well.

Contrary to YoJeff's opinion, I don't assume too much... I am willing to bet the overwhelming majority of people that fully understand what is happening would opt for the dust to be treated secondary in a trade. This will always be the preferred method when compared to inaccurately averaged trades where the DUST represents the bought price of the FULL trade.

As stated in an earlier post, DUST can result in DCA's, lower profits, some sales that include DUST are misrepresented as a gain in the sale log when they actually trade at a loss. Not to mention this all contributes to an overall inaccurate sales log and inconsistent trade history in the sales log as compared to the exchange.

Again, this is an issue... This should not be dismissed.

A main objective of PT should be gains, not losses, and accuracy of the trade, period! - A sale should not take place at a loss because an order that has a penny's worth of dust with a lower bought price satisfies the condition for a sell trigger. This very situation has happened to me several times resulting in a loss on the whole trade. This happens because the dust bought price was low enough, by comparison to the actual bought price of the whole trade. Each time, the order was bought, included DUST which had a low enough bought price that it satisfied an immediate sell order, then sold quickly at a loss. Then the whole order was re-bought at close to the sell price because the argument for the original buy was still valid. But the icing on the cake is that the sales log showed the trade sold at a handsome profit because the low dust price was the bought price for the whole order.

Again, this is an issue and should not be dismissed.

This section is supposed to be for "issues" with PT. This is an issue that I am trying to "help" resolve. I believe Taniman understands this to be the case, so I am simply trying to include examples of where and when this happens to help provide feedback for a solution.

I find it quite dismissive of you to simply make a couple of close-minded statements, then close this thread. This is an issue that I and others have recognized and it would seem that the Devs would want to take a close look at this to make PT the best it can be.

This is not a rant, this is not a support question, this is an issue and should remain open till there is a viable solution, in my opinion.

I'm only trying to help, I appreciate all the hard work that goes into this project, and as a faithful user and advocate, I simply want to help the community and the developer make PT the best it can be.

@YoJeff ... Your position and reply to this issue seems rather provoking... You seem to have taken offense and perhaps it is you that has not given the gravity of this issue enough consideration before closing both your mind and the issue... I respectfully encourage you to reconsider both.

Once again, this post is to HELP resolve an issue. The information and images are to offer support in developing a resolution.

Sincerely...

**Comment by YoJeff** [comment]
I assure you I do not take offense.  I am tasked with keeping these issues current, and proactively (and perhaps even aggressively) closing issues which are not issues, will not be fixed, are not on the roadmap, are duplicates, are support questions, a myriad of requests that do not belong on here.

I -am- a programmer.  I understand the complexities of this problem.   You did not address any of the edge issues I described.   The saying "the devil is in the details" rings true for virtually every development project.

When you buy a new coin, you own the total purchase price + existing balance (dust or not).  The exchange doesn't separate out dust.   You don't see a sub-balance of dust.   The exchange reports your total balance, the bot uses the total balance.

Your proposal requests we ignore dust, that we put coins into two buckets and only trade a portion of it.  Well, what happens when that coin is no longer dust?   We have the exact same problem just with a bigger percentage.

**Comment by taniman** [comment]
@TheBigShort111  I am gonna need. your complete log file, your trading files and screenshots from your exchange again. And also I assume you are on version 1.2.6.20?

**Comment by TheBigShort111** [comment]
@Taniman, thank you for reopening this... I was penning a reply to YoJeff, but will delete that. I appreciate you taking the time to look at this.

Yes, I am on 1.2.6.20...

Example of higher bought price as the result of dust, indicating lower than actual profit percentage which could result in a DCA instead of a quick sale at profit.
![screenshot 2018-02-27 12 03 02](https://user-images.githubusercontent.com/36313988/36759181-8a3db944-1be4-11e8-8772-411fb4f5e9e1.png)
![screenshot 2018-02-27 12 03 17](https://user-images.githubusercontent.com/36313988/36759193-97877360-1be4-11e8-9c63-51c18496e684.png)

Example of opposite - Lower bought price as the result of dust, indicating a false profit when an actual loss on the trade was the result. After the sale, an immediate rebuy occurred as the argument for the original buy was still valid...
![screenshot 2018-02-16 00 58 24](https://user-images.githubusercontent.com/36313988/36758618-fdc6682c-1be2-11e8-9652-93596fcffc37.png)
![screenshot 2018-02-27 17 29 16](https://user-images.githubusercontent.com/36313988/36759433-5e57a960-1be5-11e8-8f0c-43faae97dbbc.png)

Another example of sale at a loss, then immediate rebuy, sales log inaccurately reflects 1.64% gain...
![screenshot 2018-02-16 01 04 00](https://user-images.githubusercontent.com/36313988/36758990-fa672f3a-1be3-11e8-9361-7ab53a939c41.png)
![screenshot 2018-02-27 17 29 16](https://user-images.githubusercontent.com/36313988/36759032-17062984-1be4-11e8-8b45-d04bafde9c5e.png)

This trade (with dust) did not result in a loss, however it showed a 4.18% ($2.09) gain in the sales log, when in reality the sale was falsely triggered because it appeared to PT to meet my sell criteria. As a result of the artificial low bought price of the dust... profit was only $0.45 (I have subsequently edited the log to reflect the correct amount.
![screenshot 2018-02-16 01 19 58](https://user-images.githubusercontent.com/36313988/36759587-e778175c-1be5-11e8-8c47-5c65f07d0512.png)
![screenshot 2018-02-16 01 21 30](https://user-images.githubusercontent.com/36313988/36759597-ee0bb164-1be5-11e8-9ed4-cb72e4983709.png)

Another sale at a loss: ETH bought at $933.82, (but shows a bought price of $881.38 because of dust) - It immediately sold for $929.39 because PT calculated that it was trading for a nice profit - Sales log shows 5.24% gain in around 30 seconds for a $2.59 profit, when in actuality the trade sold at a $0.26 loss, then immediately re-bought because the original argument for the buy was still valid.
![screenshot 2018-02-16 01 17 19](https://user-images.githubusercontent.com/36313988/36760012-48394768-1be7-11e8-9f9e-327b6bdd61bc.png)
![screenshot 2018-02-27 17 55 27](https://user-images.githubusercontent.com/36313988/36760055-6eab0fc6-1be7-11e8-8055-da4034b96ece.png)

Here is another one (with the dust issue)... PT shows a bought price of $10,408.73683544, and a sale at $10,735.01 with a PT sales log profit of 2.93% ($1.55), however, as in the earlier examples, the bought price was taken from the dust that was much lower than the actual order... The actual bought price of the order was $10,751.31, resulting in a loss of $.05 rather than the indicated gain of $1.55 - Then as in previous examples, the argument was still valid for the original purchase, so there was an immediate rebuy.
![screenshot 2018-02-18 23 42 10](https://user-images.githubusercontent.com/36313988/36760353-74fd4b68-1be8-11e8-91cd-76d1836244af.png)
![screenshot 2018-02-18 23 43 23](https://user-images.githubusercontent.com/36313988/36760361-7d0a0c2e-1be8-11e8-8a83-08f6eac291ea.png)

Here's the info your requested... As mentioned earlier, I did go in and edit one or two of the transactions (I believe one was a BTC one), so all but one or two should show the unedited data... Also, my trading files have been recently changed, so they do not reflect the settings I had at the time of earlier trades. I'm assuming when you say you need my trading files, you mean the DCA, Indicator, and Pairs files, correct?

[ProfitTrailerData.txt](https://github.com/taniman/profit-trailer/files/1765061/ProfitTrailerData.txt)
[DCA.properties.txt](https://github.com/taniman/profit-trailer/files/1765075/DCA.properties.txt)
[INDICATORS.properties.txt](https://github.com/taniman/profit-trailer/files/1765076/INDICATORS.properties.txt)
[PAIRS.properties.txt](https://github.com/taniman/profit-trailer/files/1765077/PAIRS.properties.txt)

I hope this helps you figure out a solution for this ongoing issue. I encourage others that may read this (especially those that were affected by immediate rebuys, or those seeing an occasional higher than normal profit on a trade here and there to take a closer look) - Compare your sales log data with the trade history on your exchange.

If you need any more information, feel free to ask.

Thanks a bunch for looking into this. Keep up the great work!

**Comment by taniman** [comment]
quick note: I am certain (I did not watch the video yet) crypto gnome also talked about affiliate commission on his account, which is different than dust and is something that the bot cannot handle.

**Comment by TheBigShort111** [comment]
In a broad sense, the commissions are dust as well since they are insignificant fractions of a coin that is too small to trade otherwise... Regardless, the exchange treats the commissions with the method I proposed... They do not separate it, they simply will trade it along with an order at whatever price the order is executed at. In other words, the dust always takes a back seat in the trade and its original bought cost is not factored in with any trade it is used in.

So, if I have dust, whether from a previous sale, or from affiliate earnings, that amount "dust" is generally used in a new trade and is added into the mix at the price of the new trade, so if that insignificant amount of coin takes on a profit or loss, it is on an amount that otherwise was too small to trade to being with. Again, we are talking about trivial amounts here that are affecting the main trade itself in nothing but a negative way. So many times this is fractions of a cent that is potentially costing dollars in a trade - It simply does not make sense to use the dust (of affiliate dust) as the base price for a trade. The exchange does not do it that way under any circumstance, neither should PT. There is no upside to using the method PT uses, and nothing but upside to the way the exchange handles it.

If the affiliate earnings were high enough that they were no longer considered dust, then even still, the exchange treats the larger amount the exact same way... It can now either be traded independently as a single order on its own, or combined into another trade at the then prevailing price of the coin.

In a nutshell, affiliate earnings, trade dust, or any amount too small to trade independently is used at the then prevailing buy price, regardless of the original bought price of such dust (or affiliate commission). It is as straight-forward as can be. Again, it is how the exchange deals with it, and how PT should deal with it.

You would be hard pressed to find an argument for salvaging a fraction, of a fraction, of a fraction gain or loss for such a trivial amount, when the downside potential can be in the form of dollars, tens of dollars, hundreds of dollars, or even thousands (or tens of thousands) depending on the trade size. That is an irrefutable fact. Certainly there is good reason the exchange handles the trivial amounts in the fashion I have outlined... I have never once felt slighted by an exchange for having dust (or non-trade worthy amounts) of a coin, or using such amounts in a trade with any regard to the bought price of the dust itself... I'm just happy when it is able to be used in a trade, whatever the trivial/insignificant amount is worth (which is usually next to nothing)

One last point... I think it is great that PT will incorporate dust in a trade, however it should be treated with the same attitude as it does on an exchange. It is great that PT will trade with it, but also keep in mind that usually after trading with dust, new dust is often created, which creates a vicious circle. So PT does not eliminate dust on exchanges where dust exists as the result of rounding, or affiliate commissions. Frankly speaking, this should be a non-issue, and can be resolved by adopting the rules/methods by which the exchange uses to trade such trivial amounts.

Thanks for taking the time and putting forth the effort to resolve this issue. it is very much appreciated.

Keep up the great work.

**Comment by TheBigShort111** [comment]
Okay, I can finally show you an example of an buy that INSTANTLY went into DCA as the result of old DUST being used as the bought price... This is a serious issue... The effect of the dust is becoming a real issue when it is used in a pair that has not traded for a while.

So, here the trade went "immediately" into DCA becasue the dust bought price was high enough to trigger an immediate DCA! - You will see that the "actual" bought price not only would not have triggered a DCA, but was actually profitable right from the get go.

![screenshot 2018-02-28 19 23 20](https://user-images.githubusercontent.com/36313988/36820776-9f052b8a-1cbd-11e8-8e1a-ecaff6b0fbe0.png)
![screenshot 2018-02-28 19 23 42](https://user-images.githubusercontent.com/36313988/36820779-a457207a-1cbd-11e8-988e-f7ea6ac1dfde.png)

Here is the result after I manually set the price in hotconfig:

![screenshot 2018-02-28 19 25 33](https://user-images.githubusercontent.com/36313988/36820784-a9de048c-1cbd-11e8-8849-0a36927dee93.png)

Oh, and just before this happened, I had another buy for NEO where I had to go in and manually change the price before it DCA'd... Unfortunately I didn't have time to take a quick picture before changing the price in hotconfig.

Frankly speaking, I'm starting to fear running PT without constant monitoring... I'm willing to bet many people are holding unnecessary bags, and have funds tied up in DCA's as the result of this flaw. Some people just run PT without closely monitoring it... And they accept the results as indicated. That said, I'm willing to bet that PT would be performing much better than most people are seeing once this dust issue is finally resolved.

**Comment by taniman** [comment]
I can agree with you that there is something wrong with picking the correct price. 
But one thing you need to understand this logic is not designed because of possible dust. 
There are greater logics in place and it's due to partial sells and DCA.

I appreciate your debugging, but a lot of your debugging also comes with a lot of extra text making it hard for me to debug. 

You should try keeping it to facts in the future. 

tl;dr;

I will not be building a way to skip dusts for this issue. But I think I figured out the problem and will however be posting a fix for that.
